src/zenml/models/api_key_models.py

```diff
@@ -0,0 +1,296 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Models representing API keys."""
+
+from datetime import datetime, timedelta
+from typing import TYPE_CHECKING, ClassVar, List, Optional, Type, Union
+from uuid import UUID
+
+from passlib.context import CryptContext
+from pydantic import BaseModel, Field
+
+from zenml.constants import ZENML_API_KEY_PREFIX
+from zenml.models.base_models import (
+    BaseRequestModel,
+    BaseResponseModel,
+    update_model,
+)
+from zenml.models.constants import STR_FIELD_MAX_LENGTH
+from zenml.models.filter_models import BaseFilterModel
+from zenml.utils.string_utils import b64_decode, b64_encode
+
+if TYPE_CHECKING:
+    from sqlmodel.sql.expression import Select, SelectOfScalar
+
+    from zenml.models.filter_models import AnySchema
+    from zenml.models.service_account_models import ServiceAccountResponseModel
+
+# ---- #
+# BASE #
+# ---- #
+
+
+class APIKey(BaseModel):
+    """Encoded model for API keys."""
+
+    id: UUID
+    key: str
+
+    @classmethod
+    def decode_api_key(cls, encoded_key: str) -> "APIKey":
+        """Decodes an API key from a base64 string.
+
+        Args:
+            encoded_key: The encoded API key.
+
+        Returns:
+            The decoded API key.
+
+        Raises:
+            ValueError: If the key is not valid.
+        """
+        if encoded_key.startswith(ZENML_API_KEY_PREFIX):
+            encoded_key = encoded_key[len(ZENML_API_KEY_PREFIX) :]
+        try:
+            json_key = b64_decode(encoded_key)
+            return cls.parse_raw(json_key)
+        except Exception:
+            raise ValueError("Invalid API key.")
+
+    def encode(self) -> str:
+        """Encodes the API key in a base64 string that includes the key ID and prefix.
+
+        Returns:
+            The encoded API key.
+        """
+        encoded_key = b64_encode(self.json())
+        return f"{ZENML_API_KEY_PREFIX}{encoded_key}"
```

Comment:
This is also something stripe does. Separating the prefix from the rest of the key makes it easier to spot and read

```suggestion
        return f"{ZENML_API_KEY_PREFIX}_{encoded_key}"
```

Reply:
done (in the `ZENML_API_KEY_PREFIX` definition)